### ISSUE TITLE
udev links break with newer systemd versions

### ISSUE THREAD

**Issue by atkinsam** [LINBIT/drbd-utils#44]
DRBD's udev rules don't work with newer systemd versions when the machine's hostname changes during boot, as it does in many VMs and cloud hosts. Starting with systemd version 242, the new [`ProtectHostname`](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#ProtectHostname=) sandbox setting is enabled by default in the `systemd-udevd` service, applying to any scripts run within udev rules (see [systemd-udevd unit file](https://github.com/systemd/systemd/blob/0068131bf5c777618e2df7c4769fc6844de2359e/units/systemd-udevd.service.in#L32) and [changelog](https://github.com/systemd/systemd/blob/0068131bf5c777618e2df7c4769fc6844de2359e/NEWS#L7599)). This is impacting DRBD's udev links: https://github.com/LINBIT/drbd-utils/blob/1caa04161811e085c7a1c4847c07eb663cf85e3f/scripts/drbd.rules.in#L6

When the udev rule executes, drbdadm sees the host's original (invalid) hostname, compares it to the hostname in the resource config, and fails:

```
(udev-worker)[93198]: drbd0: Starting '/usr/local/sbin/drbdadm sh-udev minor-0'
(udev-worker)[93198]: drbd0: '/usr/local/sbin/drbdadm sh-udev minor-0'(out) ''minor-0' not defined in your config (for this host).'
(udev-worker)[93198]: drbd0: Process '/usr/local/sbin/drbdadm sh-udev minor-0' failed with exit code 1.
```

And the `/dev/drbd/by-disk` and `/dev/drbd/by-res` symlinks never get created.

I'm not sure what the right solution to this is. As far as I can tell, there is no way to scope these sandbox rules to individual udev rules or scripts. Disabling ProtectHostname on `systemd-udevd` via a systemd override solves this problem. Restarting the `systemd-udevd` service also solves this problem, as the hostname in the UTS namespace is refreshed. Neither of those seem like tenable solutions to me.

**Comment by raltnoeder** [comment]
Or rather, newer systemd versions don't work with DRBD's udev rules, because they invented new dirty magic tricks to avoid the fallout from slightly older dirty magic tricks invented by the VM and cloud crowd. I am getting kind of used to things constantly breaking as soon as virtualization, cloud, systemd or udev are involved, but still, fixing it will probably take a while, because we might have to set up an environment where we can recreate the breakage to check whether any workarounds that we can come up with will work.

**Comment by WanzenBug** [comment]
We are working on resolving the issue by changing the udev rules to use a new `drbdsetup` command. This is actually a much cleaner solution in any case, as that takes the actual configuration in the kernel, compared to stuff in the config files.

There is less chance of breakage caused by invalid config files, mismatched configuration or, as with this issue, mismatched hostnames.

**Comment by JoelColledge** [comment]
Fixed in drbd-utils v9.28.0. From the announcement message:

> udev: drbdsetup got a new "udev" subcommand that replaces drbdadm's
  "sh-udev" command. This allows to have the actual (in kernel) state
  reflect the udev state more closely compared to drbdadm which used res
  file information. It avoids problems if the res file can not be parsed
  or problems if the hostname is not set yet (e.g., systemd-udev starts
  before the hostname is set). All in all we hope this solves udev
  related issues we saw previously.

https://lists.linbit.com/pipermail/drbd-announce/2024-May/000728.html
